src/libraries/System.Net.Mail/src/System/Net/Base64Stream.cs

```diff
@@ -199,104 +199,167 @@ internal int EncodeBytes(byte[] buffer, int offset, int count, bool dontDeferFin
+        private static int GetCodepointSize(string value, int i)
+        {
+            if (char.IsSurrogate(value[i]) && i + 1 < value.Length && char.IsSurrogatePair(value[i], value[i + 1]))
             {
+                return 2;
+            }
+            return 1;
+        }
+
+        private int AppendEncodedCodepoint(byte[] bytes, int count, bool shouldAppendSpaceToCRLF)
+        {
+            if (LineBreakNeeded(count)) {
+                AppendPadding();
                 WriteState.AppendCRLF(shouldAppendSpaceToCRLF);
             }
+            for (int i = 0; i < count; ++i)
+            {
+                ApppendEncodedByte(bytes[i]);
+            }
 
-            //now pad this thing if we need to.  Since it must be a number of bytes that is evenly divisble by 3,
-            //if there are extra bytes, pad with '=' until we have a number of bytes divisible by 3
-            switch (count % 3)
+            return count;
+        }
+
+        private int ApppendEncodedByte(byte b)
+        {
+            Debug.Assert(0 <= WriteState.Padding && WriteState.Padding <= 2, "paddind was not in range [0,2]");
+
+            // Base64 encoding transforms a group of 3 bytes into a group of 4 Base64 characters
+            switch (WriteState.Padding)
             {
-                case 2: //One character padding needed
-                    WriteState.Append(Base64EncodeMap[(buffer[cur] & 0xFC) >> 2]);
-                    WriteState.Append(Base64EncodeMap[((buffer[cur] & 0x03) << 4) | ((buffer[cur + 1] & 0xf0) >> 4)]);
-                    if (dontDeferFinalBytes)
-                    {
-                        WriteState.Append(Base64EncodeMap[((buffer[cur + 1] & 0x0f) << 2)]);
-                        WriteState.Append(Base64EncodeMap[64]);
-                        WriteState.Padding = 0;
-                    }
-                    else
-                    {
-                        WriteState.LastBits = (byte)((buffer[cur + 1] & 0x0F) << 2);
-                        WriteState.Padding = 1;
-                    }
-                    cur += 2;
+                case 0: // Add first byte of 3
+                    WriteState.Append(Base64EncodeMap[(b & 0xfc) >> 2]);
+                    WriteState.LastBits = (byte)((b & 0x03) << 4);
+                    WriteState.Padding = 2;
+                    return 1;
+                case 2: // Add second byte of 3
+                    WriteState.Append(Base64EncodeMap[WriteState.LastBits | ((b & 0xf0) >> 4)]);
+                    WriteState.LastBits = (byte)((b & 0x0f) << 2);
+                    WriteState.Padding = 1;
+                    return 1;
+                case 1: // Add third byte of 3
+                    WriteState.Append(Base64EncodeMap[WriteState.LastBits | ((b & 0xc0) >> 6)]);
+                    WriteState.Append(Base64EncodeMap[(b & 0x3f)]);
+                    WriteState.Padding = 0;
+                    return 1;
```

Comment:
`default` case is missing. Is it intentional?

Reply:
Added default, but actually padding should be 0, 1 or 2 by design
https://github.com/dotnet/runtime/blob/e8c55d73aa809b072b780f38541ba9ccb7e28667/src/libraries/System.Net.Mail/src/System/Net/Mime/Base64Encoder.cs#L126

Reply:
I'm not insisting on adding `default` if it's by design. Just asking :)
If it's intentional in the given case I then would recommend to keep it as it is (no `default`) or throw an exception (e.g. `AgrumentOutOfRangeException`). Having a blank default doesn't add any value.

Reply:
I've consulted @antonfirsov about AgrumentOutOfRangeException here and the idea was that because it's an internal logic, we should rather use Debug.Assert than an exception. I will remove empty defaults then

Reply:
Used Debug.Fail in defaults so the logic would be more clear